Add memory functions implementation for AArch64

thiru-mcw opened the pull request:
- Based on information from https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/issues/223 ticket Implemented memory functions for AArch64 architecture in C. 

Validated by 
https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/13155525706

eukarpov commented:
LGTM as a temporary fix and the work should be resumed later with implementing these functions using aarch64 instructions for performance optimization.
aarch64-cygwin-wip-v3 will be merged to woarm64 after adding "Hello Cygwin" to CI. Then this PR can be merged directly
to woarm64.

Blackhex reviewed (approved):
Build verified by this workflow run https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/12650121489, approving. Please wait with the merge for now.

eukarpov commented:
> Build verified by this workflow run https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/12650121489, approving. Please wait with the merge for now.

It is not enough. We need to validate hello-cygwin and revert changes related to memory operations in previous patches.

thiru-mcw commented:
Examples for testing mem-functions:

Test example: memchr()
```
#include <stdio.h>
#include <string.h>
void *memchr(const void *s, int c, size_t n) __declspec(dllimport);
int main()
{
    char buffer[10];
    for (int i = 0; i < 10; i++)
    {
        buffer[i] = '.';
    }
    buffer[5] = 'X';
    char *found = (char *)memchr(buffer, 'X', sizeof(buffer));
    if (found != NULL && (found - buffer) == 5)
    {
        puts("memchr works");
    }
    else
    {
        puts("memchr failed");
    }
    return 100;
}
```

Test example: memset()
```
#include <stdio.h>
#include <stddef.h>
#include <string.h>
void *memset(void *dest, int c, size_t n) __declspec(dllimport);

int main()
{
    char buffer[5];
    for (int i = 0; i < 5; i++)
    {
        buffer[i] = '.';
    }
    memset(buffer, '*', sizeof(buffer));

    int success = 1;
    for (int i = 0; i < sizeof(buffer); i++)
    {
        if (buffer[i] != '*')
        {
            success = 0;
            break;
        }
    }
    if (success)
    {
        puts("memset works");
    }
    else
    {
        puts("memset failed");
    }
    return 100;
}

```
Test example: mempcpy()
```
#include <stdio.h>
#include <string.h>
void *mempcpy(void *dest, const void *src, size_t n) __declspec(dllimport);
int main() 
{
    char a[3];
    char b[3];
    for (int i = 0; i < 3; i++) {
        a[i] = 'a';
        b[i] = 'b'; 
    }
    char* end_of_b = mempcpy(b, a, sizeof(a));
    int success = 1;
    for (int i = 0; i < sizeof(b); i++) {
        if (b[i] != 'a') {
            success = 0;
            break;
        }
    }
    if (success) {
        puts("mempcpy works");
    }
    else
    {
        puts("mempcpy failed");
    }
    return 100;
}

```

eukarpov commented:
btw why memset, memchr and mempcpy cannot be used from here?
newlib/libc/string/memset.c

Blackhex commented on newlib/libc/machine/aarch64/memcpy.S:
```diff
@@ -56,6 +56,7 @@
    The loop tail is handled by always copying 64 bytes from the end.
 */
 
+ENTRY_ALIAS (memmove)
```
Any reasoning for this change? `memmove` cares about overlapping memory while `memcpy` does not. Couldn't this lead to unpredictable issues when `memcpy` implementation is used for `memmove`?

Blackhex replied:
I see it now, it reverts the previous workaround.

Blackhex replied:
On a second thought, did we test if the current `memcpy` implementation behaves correctly in scenarios where `memmove` needs to be used?

thiru-mcw replied:
Thanks for bringing this up!
The current `memcpy` implementation performs poorly in overlap scenarios where `memmove` should be used. 
I'll make sure to refactor it, aligning it with the newlib implementation and properly handling of overlaps using memmove.

Edit: Refactoring Done, overlap cases are handled with memmove. 

https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/blob/8e071e00c2c5ed14b7aad2a352fc2c98f150fe5d/winsup/cygwin/aarch64/memcpy.c#L56-L59

Blackhex commented on winsup/cygwin/aarch64/memchr.S:
This file was added as a workaround. It should no longer be needed, does it? Or shouln't it be filled with a proper implementation? The same applies to other empty `*.S` files.

thiru-mcw replied:
Thank you for pointing that out! 
The commit hashes on the `woarm64` branch were altered last night. I’ve rebased the PR on the current `woarm64` branch.

Blackhex reviewed (changes requested):
Left some comments. My biggest concern is why the empty `*.S` files stayed unreverted and this https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/pull/3#issuecomment-2616557988 comment as well.

thiru-mcw commented:
> btw why memset, memchr and mempcpy cannot be used from here?
> newlib/libc/string/memset.c

@eukarpov Thanks for the suggestion! 
I tested the implementations of `memset`, `memchr`, and `mempcpy` from the `newlib/libc/string` directory, and the sample examples passed successfully during local compilation from scratch. 

Since these implementations closely align with the existing Cygwin codebase, I have refactored the PR to incorporate them from the newlib directory.

eukarpov commented:
> > btw why memset, memchr and mempcpy cannot be used from here?
> > newlib/libc/string/memset.c
> 
> @eukarpov Thanks for the suggestion! I tested the implementations of `memset`, `memchr`, and `mempcpy` from the `newlib/libc/string` directory, and the sample examples passed successfully during local compilation from scratch.
> 
> Since these implementations closely align with the existing Cygwin codebase, I have refactored the PR to incorporate them from the newlib directory.

Good! Please remove previously introduced implementation. Tests are not needed in this case. It should be handled by cygwin unit tests.

eukarpov commented on winsup/cygwin/aarch64/memchr.c:
```diff
@@ -0,0 +1,126 @@
+/*
```
why mem* are not reused from newlib/libc/string as it was discussed and tested?

thiru-mcw replied:
The same Implementations are reused here, I have updated the branch 5hrs ago, can you please check again. Thanks

eukarpov replied:
Could it be direct references to the implementation without copying files?

Blackhex replied:
I'd also prefer either using `newlib` implementations directly (via linking, if possible) or referencing them in the build scripts (if possible).  Either way, it would be best to discuss with Cygwin maintainers what would be most acceptable way for them.

thiru-mcw replied:
Sure, Let me check by linking `newlib` Implementations directly and update on this.

eukarpov commented on winsup/cygwin/mm/malloc_wrapper.cc:
```diff
@@ -50,13 +50,27 @@ import_address (void *imp)
 {
   __try
     {
+#if defined(__aarch64__)
```
please rebase with the latest changes

eukarpov replied:
the issue with cygwin daily rebase again. I will fix it.

eukarpov replied:
It is fixed. However rebased is still needed.

thiru-mcw replied:
Sure, will rebase it now.

eukarpov commented on winsup/cygwin/Makefile.am:
```diff
@@ -63,14 +63,16 @@ TARGET_FILES= \
 	aarch64/bcopy.S \
-	aarch64/memchr.S \
+	$(NEWLIB_STRING_DIR)/memchr.c \
 	aarch64/memcpy.c \
```
why newlib/libc/string/memcpy.c is not reused?

thiru-mcw replied:
@eukarpov The `memcpy` implementation in newlib directory doesn't handle the overlap scenarios where `memmove` should be used.

Thus, the same file is reused here in winsup with added handling for the edge case of memory overlap.  

https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/blob/8e071e00c2c5ed14b7aad2a352fc2c98f150fe5d/winsup/cygwin/aarch64/memcpy.c#L56-L59

eukarpov replied:
What about patching the original file at the original place?

thiru-mcw replied:
Heard that the `newlib` is maintained by a different set of maintainers, so any change there would be hard to get upstreamed. That's why the fix was applied in `winsup` instead.

eukarpov replied:
It might be that memcpy.S on aarch64 implementation will be implemented by upstreaming time. Now for simplicity and maintainability let's make changes at the original place.

eukarpov replied:
Actually it looks like memcpy.S is reverted back to initial implementation in this PR. Why memcpy.c is needed?

thiru-mcw replied:
With the help of Radek, got information from the maintainers. The assembly implementation in `newlib` is designed for the SysV ABI, while the C implementations serve as a fallback mechanism, used only when no assembly implementation is used.

eukarpov replied:
Thanks for clarification. Based on registers usage defined here https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/blob/ab3ffbc3848b754a3aa84ef7d94295db74410c9d/newlib/libc/machine/aarch64/memcpy.S#L19

it might be compatible with Arm64 ABI. Could we validate if it actually works?
It does not have SEH wrapper, however it might be not needed for such a implementation.

thiru-mcw replied:
To use this [newlib-cygwin/newlib/libc/machine/aarch64/memcpy.S](https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/blob/ab3ffbc3848b754a3aa84ef7d94295db74410c9d/newlib/libc/machine/aarch64/memcpy.S#L19) Implementation, 
The following line has to be removed as it causes multiple definition of `memmove` and breaks the toolchain build.
https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/blob/ab3ffbc3848b754a3aa84ef7d94295db74410c9d/newlib/libc/machine/aarch64/memcpy.S#L59

Removing above line and re-building toolchain, its successful and the `memcpy` works as expected.

eukarpov replied:
Maybe rebased was missed, it should work without removing the line.
Please apply the same changes for the rest of files in winsup/cygwin/aarch64/
bcopy.S and etc to revert changes in that folder fully.

eukarpov replied:
and $(NEWLIB_STRING_DIR)/memmove.c should be removed

Blackhex replied:
AFAIK, this will make \`memove\` implementation working incorrectly in memory overlapping scenarios. Or is my information already outdated?

eukarpov replied:
Could you please give more context why?

eukarpov replied:
Based on the description in memcpy.S

```
This implementation handles overlaps and supports both memcpy and memmove
   from a single entry point.
```

thiru-mcw replied:
The memcpy.S handles overlaps with backward copying.

https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/blob/ab3ffbc3848b754a3aa84ef7d94295db74410c9d/newlib/libc/machine/aarch64/memcpy.S#L201-L202

@eukarpov I'll try removing `$(NEWLIB_STRING_DIR)/memmove.c` and check if toolchain builds without multiple definition Error.

thiru-mcw replied:
Build validated in [Build toolchain variants #1672 ](https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/13155525706).

eukarpov commented on winsup/cygwin/Makefile.am:
```diff
@@ -63,18 +63,20 @@ TARGET_FILES= \
-	aarch64/wmempcpy.S
+	$(NEWLIB_STRING_DIR)/bcopy.c \
+	$(NEWLIB_STRING_DIR)/memchr.c \
+	$(NEWLIB_MACHINE_DIR)/memcpy.S \
```
Thanks for the update. I think the PR is almost ready.
I expect that memcpy.S is not needed here and perhaps it is declared here
newlib/libc/machine/aarch64/Makefile.inc

thiru-mcw replied:
Done, thanks

thiru-mcw committed:
Add mem-funcs implementation for aarch64

Linking memory functions from newlib/libc/string in winsup makefile for AArch64 target.

eukarpov reviewed (approved):
LGTM. Thanks for the contribution.

Blackhex reviewed (approved):
I can accept the changes for now as they seem reasonable from technical standpoint. In the future, we need to align with Cygwin maintainers whether such implementation would be acceptable.
